eureka/lib/tracing_niriss.py

```diff
@@ -0,0 +1,410 @@
+from astropy.modeling.models import Moffat1D
+
+__all__ = ['image_filtering', 'simplify_niriss_img',
+           'mask_method_edges', 'mask_method_ears', 'f277_mask',
+           'ref_file']
+
+
+def image_filtering(img, radius=1, gf=4):
+    """
+    Does some simple image processing to isolate where the
+    spectra are located on the detector. This routine is
+    optimized for NIRISS S2 processed data and the F277W filter.
+
+    Parameters
+    ----------
+    img : np.ndarray
+       2D image array.
+    radius : np.float, optional
+       Default is 1.
+    gf : np.float, optional
+       The standard deviation by which to Gaussian
+       smooth the image. Default is 4.
+
+    Returns
+    -------
+    img_mask : np.ndarray
+       A mask for the image that isolates where the spectral
+       orders are.
+    """
+    mask = filters.rank.maximum(img/np.nanmax(img),
+                                disk(radius=radius))
+    mask = np.array(mask, dtype=bool)
+
+    # applies the mask to the main frame
+    data = img*mask
+    g = gaussian_filter(data, gf)
+    g[g>4] = 10000
+    edges = filters.sobel(g)
+    edges[edges>0] = 10
+
+    # turns edge array into a boolean array
+    edges = (edges-np.nanmax(edges)) * -1
+    z = feature.canny(edges)
+
+    return z, g
+
+
+def simplify_niriss_img(data):
+    """
+    Creates an image to map out where the orders are in
+    the NIRISS data.
+
+    Parameters
+    ----------
+    data : np.array
+       A 3D array of all frames to calculate the
+       maximum frame.
+
+    Returns
+    -------
+    g : np.ndarray
+       A 2D array that marks where the NIRISS first
+       and second orders are.
+    """
+    perc  = np.nanmax(data, axis=0)
+    # creates data img mask
+    z,g = image_filtering(perc)
+    return g
+
+
+def f277_mask(f277, radius=1, gf=4):
+    """
+    Marks the overlap region in the f277w filter image.
+
+    Parameters
+    ----------
+    data : object
+
+    Returns
+    -------
+    mask : np.ndarray
+       2D mask for the f277w filter.
+    mid : np.ndarray
+       (x,y) anchors for where the overlap region is located.
+    """
+    img = np.nanmax(f277, axis=(0,1))
+    mask, _ = image_filtering(img[:150,:500], radius, gf)
+    mid = np.zeros((mask.shape[1], 2),dtype=int)
+    new_mask = np.zeros(img.shape)
+
+    for i in range(mask.shape[1]):
+        inds = np.where(mask[:,i]==True)[0]
+        if len(inds) > 1:
+            new_mask[inds[1]:inds[-2], i] = True
+            mid[i] = np.array([i, (inds[1]+inds[-2])/2])
+
+    q = ((mid[:,0]<420) & (mid[:,1]>0) & (mid[:,0] > 0))
+
+    f277_img = new_mask
+
+    return new_mask, mid[q]
+
+
+def mask_method_edges(data, radius=1, gf=4,
+                      save=False,
+                      outdir=None):
+    """
+    There are some hard-coded numbers in here right now. The idea
+    is that once we know what the real data looks like, nobody will
+    have to actually call this function and we'll provide a CSV
```

Comment:
CSV isn't great as it's hard to package together with the rest of the code, but I understand the general idea